decayedlog.go

```diff
@@ -0,0 +1,442 @@
+// database entries.
+func (d *DecayedLog) Start() error {
+	if !atomic.CompareAndSwapInt32(&d.started, 0, 1) {
+		return nil
+	}
+
+	// Open the boltdb for use.
+	var err error
+	if d.db, err = bolt.Open(d.dbPath, dbPermissions, nil); err != nil {
+		return fmt.Errorf("Could not open boltdb: %v", err)
+	}
+
+	// Initialize the primary buckets used by the decayed log.
+	if err := d.initBuckets(); err != nil {
+		return err
+	}
+
+	// Start garbage collector.
+	if d.notifier != nil {
+		epochClient, err := d.notifier.RegisterBlockEpochNtfn()
+		if err != nil {
+			return fmt.Errorf("Unable to register for epoch "+
+				"notifications: %v", err)
+		}
+
+		d.wg.Add(1)
+		go d.garbageCollector(epochClient)
+	}
+
+	return nil
+}
+
+// initBuckets initializes the primary buckets used by the decayed log, namely
+// the shared hash bucket, and batch replay
+func (d *DecayedLog) initBuckets() error {
+	return d.db.Update(func(tx *bolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(sharedHashBucket)
+		if err != nil {
+			return ErrDecayedLogInit
+		}
+
+		_, err = tx.CreateBucketIfNotExists(batchReplayBucket)
+		if err != nil {
+			return ErrDecayedLogInit
+		}
+
+		return nil
+	})
+}
+
+// Stop halts the garbage collector and closes boltdb.
+func (d *DecayedLog) Stop() error {
+	if !atomic.CompareAndSwapInt32(&d.stopped, 0, 1) {
+		return nil
+	}
+
+	// Stop garbage collector.
+	close(d.quit)
+
+	d.wg.Wait()
+
+	// Close boltdb.
+	d.db.Close()
+
+	return nil
+}
+
+// garbageCollector deletes entries from sharedHashBucket whose expiry height
+// has already past. This function MUST be run as a goroutine.
+func (d *DecayedLog) garbageCollector(epochClient *chainntnfs.BlockEpochEvent) {
+	defer d.wg.Done()
+	defer epochClient.Cancel()
+
+	for {
+		select {
+		case epoch, ok := <-epochClient.Epochs:
+			if !ok {
+				// Block epoch was canceled, shutting down.
+				sphxLog.Infof("Block epoch canceled, " +
+					"decaying hash log shutting down")
+				return
+			}
+
+			// Perform a bout of garbage collection using the
+			// epoch's block height.
+			height := uint32(epoch.Height)
+			if err := d.gcExpiredHashes(height); err != nil {
+				sphxLog.Errorf("unable to expire hashes at "+
+					"height=%d", height)
+			}
+
+		case <-d.quit:
+			// Received shutdown request.
+			sphxLog.Infof("Decaying hash log received " +
+				"shutdown request")
+			return
+		}
+	}
+}
+
+// gcExpiredHashes purges the decaying log of all entries whose CLTV expires
+// below the provided height.
+func (d *DecayedLog) gcExpiredHashes(height uint32) error {
+	return d.db.Batch(func(tx *bolt.Tx) error {
+		// Grab the shared hash bucket
+		sharedHashes := tx.Bucket(sharedHashBucket)
+		if sharedHashes == nil {
+			return fmt.Errorf("sharedHashBucket " +
+				"is nil")
+		}
+
+		var expiredCltv [][]byte
+		if err := sharedHashes.ForEach(func(k, v []byte) error {
+			// Deserialize the CLTV value for this entry.
```

Comment:
In the future, we may want to modify the primary key to actually be the height. If so, then this becomes a range scan with a known termination point, rather than reading _every_ stored hash. The tradeoff is that this would be an additional index `height -> {hashes...}.

Reply:
Do you think it's worth making this garbage collect via a ticker, so that we don't rescan the entire db for each block during a resync?

Reply:
Ah, yeh I hadn't considered the resync case. Yeah a ticker may make sense in this case, either way we can follow up with a proper solution when/if we run into any performance related issues.